### ISSUE TITLE
Display order of properties of a tensor object in debugging

### ISSUE THREAD

**Issue by Jesbong** [microsoft/debugpy#1525]
Can I adjust the display order of properties of a tensor object when debugging Python in vscode? Because in deep learning, we focus more on the shape of a tensor rather than specific values. Now we want to see the value of a tensor by pulling it down a long distance. This has reduced my debugging efficiency.

**Comment by int19h** [comment]
We don't officially support this, but I believe you could author a pydevd extension to customize the representation for your type - have a look at https://github.com/microsoft/debugpy/tree/main/src/debugpy/_vendored/pydevd/pydevd_plugins/extensions. Note however that we will be moving away from pydevd in the future.

However, given that ML/AI is a particularly important scenario these days, I'd love to hear more about the nature of customizations in question, so that we can properly address this in our new debugger.

**Comment by Jesbong** [comment]
The shape(dimension) of a tensor in vscode is far away from the top.  Frequent need to expand and view results in low efficiency.
![v2-d26e166332f23874578a00156232284e_r](https://github.com/microsoft/debugpy/assets/43310393/5c52df79-7102-4786-afb3-b430acb33bb1)


And pycharm can see the type and shape without opening the variable, as shown in the figure. We are very concerned about the dimensions of tensors in ML/AI,  it's so convenient.
![pycharm](https://github.com/microsoft/debugpy/assets/43310393/51cbb0a9-571d-41aa-9858-f6ccdf489526)

**Comment by Jesbong** [comment]
I believe the following issue are all about the same thing. The shape of numpy and tensor is so important for debuger in ML/AI.
https://github.com/microsoft/debugpy/issues/1191
https://github.com/pytorch/pytorch/issues/35071
https://github.com/pytorch/pytorch/issues/63855
https://github.com/microsoft/vscode-python/pull/14244

**Comment by 54nzg** [comment]
The shape of a variable cannot be easily viewed.Whether mouse hover will be supported in the future to see the shape of the variable.

**Comment by makaspacex** [comment]
It is obvious that everyone is very satisfied with the debugging of pycharm. If there are no patents or other restrictions, you can copy the display stype and information of pycharm. If there are restrictions, ignore my comments pls.

**Comment by makaspacex** [comment]
In addition to tensor and numpy, pycharm also has good type display support for list, tuple, dict, etc.

**Comment by haduoken** [comment]
I tempory use this
https://stackoverflow.com/questions/70704619/is-it-possible-to-show-variable-shapes-and-lengths-in-vscode-python-debugger-jus

![image](https://github.com/microsoft/debugpy/assets/20755758/29246693-3bbe-4f62-b867-45895152a048)

but I strongly suggest vscode to support it natively

**Comment by ChaXxl** [comment]
I strongly suggest VS Code can show the len/shape of an countable object when debug, which PyCharm can do this, thanks to your work~~

**Comment by xuanhua** [comment]
any progress on this issue ?   vscode is so unefficient for debugging deep learning program

**Comment by qige96** [comment]
> any progress on this issue ? vscode is so unefficient for debugging deep learning program

I made a quick attempt to follow the advice

> We don't officially support this, but I believe you could author a pydevd extension to customize the representation for your type - have a look at https://github.com/microsoft/debugpy/tree/main/src/debugpy/_vendored/pydevd/pydevd_plugins/extensions. Note however that we will be moving away from pydevd in the future.

Here is how I did: find the local directory where my pydevd extensions sit, in my case is `C:\Users\ruiqi\.vscode\extensions\ms-python.debugpy-2024.10.0-win32-x64\bundled\libs\debugpy\_vendored\pydevd\pydevd_plugins\extensions\`; in the `types` sub-directory, create a new file called `pydevd_plugin_pytorch_tensor_str.py`
```python
'''
A simple example to show tensor shape on debugger
'''
from _pydevd_bundle.pydevd_extension_api import StrPresentationProvider
from .pydevd_helpers import find_mod_attr


class PyTorchTensorShapeStr:
    def can_provide(self, type_object, type_name):
        torch_tensor = find_mod_attr('torch', 'Tensor')
        return torch_tensor is not None and issubclass(type_object, torch_tensor)

    def get_str(self, val):
        return f'shape: {val.shape}, value: {val}'

import sys

if not sys.platform.startswith("java"):
    StrPresentationProvider.register(PyTorchTensorShapeStr)
```

Then, reload my vscode, and I can see the shapes

![image](https://github.com/user-attachments/assets/55fccf0a-d9db-4892-94b2-9565495fe3e0)


Hope this may help you.

**Comment by xuanhua** [comment]
@qige96  glad to hear this, I will try  👍

**Comment by xuanhua** [comment]
@qige96  Hi，sorry for this late feedback, I just tried to add a extension under debugpy like you did. 
But I got following error during lauching debugging for a python file:

```bash
Unable to load extension: pydevd_plugins.extensions.pydevd_plugin_pytorch_tensor_str
```

I also add the root directory to the PYTHONPATH, which is required by link: https://github.com/microsoft/debugpy/tree/main/src/debugpy/_vendored/pydevd/pydevd_plugins/extensions

But still the same error. And it cannot work both on my macbook pro and a container environment.  My vscode version is `debugpy-2024.12.0-darwin-arm64`

Any suggestions ?

**Comment by qige96** [comment]
> Unable to load extension: pydevd_plugins.extensions.pydevd_plugin_pytorch_tensor_str
> I also add the root directory to the PYTHONPATH, which is required by link: https://github.com/microsoft/debugpy/tree/main/src/debugpy/_vendored/pydevd/pydevd_plugins/extensions
> 
> But still the same error. And it cannot work both on my macbook pro and a container environment. My vscode version is `debugpy-2024.12.0-darwin-arm64`

Hi @xuanhua , I am not sure exactly about what was going wrong with your case, but I guess you put the plugin file in the wrong directory (due to my misleading words, sorry). You should put in `debugpy\_vendored\pydevd\pydevd_plugins\extensions\types` directory like this

![Image](https://github.com/user-attachments/assets/13c72c87-b2e6-444e-8dec-57435f3de79f)

If you put it in ` `debugpy\_vendored\pydevd\pydevd_plugins\extensions`, which is the parent directory, then you would get the error message as you mentioned.

Hope this help you.

**Comment by xuanhua** [comment]
@qige96  yes, following your suggestions, I move the my added file into `types` directory and it works now 👍

**Comment by xuanhua** [comment]
@qige96  yes, following your suggestions, I move the my added file into `types` directory and it works now 👍

**Comment by xuanhua** [comment]
@qige96 , Thanks for your solution, it maks my life really better, but one more question about these pydevd extentions,  I am often used to use docker container as the dev environment. And I found in a new container, I have to copied them to that container from host manually. Is there any better way to avoid such manual copying everytime lauching a totally new container

**Comment by qige96** [comment]
> Is there any better way to avoid such manual copying everytime lauching a totally new container

Hi @xuanhua , if they pydevd extension can merge my code into its codebase, then all problems will be gone. However, the author of Pydevd had a concern of the code in my PR to be an over-design (which indeed I admit). I am still thinking a better design to make us AI/ML people happy, but also comfort other users. But it takes time, not only for redesigning the code, but also for all other Open Source project processes, such as code review, and the author of Pydevd still has other things to do.

For now and for you, the best way could be writing a script to do all the copying/downloading jobs, and execute that script in your Dockerfile, so that everytime you launch a new container.

**Comment by fabioz** [comment]
Hey, just to note, the whole plugins idea means you shouldn't need to change the debugger code for this, you can just add it to your own code or create your own library just with that (then you don't need to worry as much about all the other use cases out there -- sure, it'd be nice to have a fix in pydevd which would fix it for all, but in that case the bar is really higher so that it doesn't break unrelated use cases).

The readme https://github.com/fabioz/PyDev.Debugger/blob/main/pydevd_plugins/extensions/README.md should explain the structure you can have without adding code to pydevd.

**Comment by qige96** [comment]
> Hey, just to note, the whole plugins idea means you shouldn't need to change the debugger code for this, you can just add it to your own code or create your own library just with that 

Hi @fabioz , thanks for the reminder and that would be a good news for us. We don't need to bother the source repository. But I just had a quick try yet it didn't work. Here is my `pydevd_plugins` folder. I am sure the parent directory of `pydevd_plugins` was in PYTHONPATH. Could you please provide some hints on what may go erong?

```
pydevd_plugins
│- __init__.py
└─extensions
       - pydevd_plugin_shape_str.py
       - __init__.py
```

```python
# __init__.py
try:
    __import__('pkg_resources').declare_namespace(__name__)
except ImportError:
    import pkgutil
    __path__ = pkgutil.extend_path(__path__, __name__)
```

```python
# pydevd_plugin_shape_str.py
import sys
sys.path.append(r'C:\Users\ruiqi\.cursor\extensions\ms-python.debugpy-2024.6.0-win32-x64\bundled\libs\debugpy\_vendored\pydevd')
print(sys.path)

from _pydevd_bundle.pydevd_extension_api import StrPresentationProvider
# from pydevd_helpers import find_mod_attr, find_class_name

def find_mod_attr(mod_name, attr):
    mod = find_cached_module(mod_name)
    if mod is None:
        return None
    return getattr(mod, attr, None)


def find_class_name(val):
    class_name = str(val.__class__)
    if class_name.find(".") != -1:
        class_name = class_name.split(".")[-1]

    elif class_name.find("'") != -1:  # does not have '.' (could be something like <type 'int'>)
        class_name = class_name[class_name.index("'") + 1 :]

    if class_name.endswith("'>"):
        class_name = class_name[:-2]

    return class_name


class SizedShapeStr:
    '''Displays the size of a Sized object before displaying its value.
    '''
    def can_provide(self, type_object, type_name):
        sized_obj = find_mod_attr('collections.abc', 'Sized')
        return sized_obj is not None and issubclass(type_object, sized_obj)

    def get_str(self, val):
        if hasattr(val, 'shape'):
            return f'shape: {val.shape}, value: {val}'
        return f'len: {len(val)}, value: {val}'

if not sys.platform.startswith("java"):
    StrPresentationProvider.register(SizedShapeStr)
```


[pydevd_plugins.zip](https://github.com/user-attachments/files/17478421/pydevd_plugins.zip)

**Comment by Windrain7** [comment]
@qige96 Thanks for your idea. But why I failed?
![Image](https://github.com/user-attachments/assets/82d59f87-b487-4d2e-92cb-5c0631036894)

**Comment by pheotry** [comment]
> > Hey, just to note, the whole plugins idea means you shouldn't need to change the debugger code for this, you can just add it to your own code or create your own library just with that嘿，请注意，整个插件的想法意味着您不需要为此更改调试器代码，您可以将其添加到您自己的代码中或使用它来创建自己的库
> 
> Hi [@fabioz](https://github.com/fabioz) , thanks for the reminder and that would be a good news for us. We don't need to bother the source repository. But I just had a quick try yet it didn't work. Here is my `pydevd_plugins` folder. I am sure the parent directory of `pydevd_plugins` was in PYTHONPATH. Could you please provide some hints on what may go erong?嗨，[@fabioz](https://github.com/fabioz)，感谢您的提醒，这对我们来说是个好消息。我们不需要打扰源存储库。但我只是快速尝试了一下，但没有用。这是我`pydevd_plugins` 文件夹。我确定 `pydevd_plugins` 的父目录在 PYTHONPATH 中。您能否提供一些提示，说明可能会发生什么？
> 
> ```
> pydevd_plugins
> │- __init__.py
> └─extensions
>        - pydevd_plugin_shape_str.py
>        - __init__.py
> ```
> 
> # __init__.py
> try:
>     __import__('pkg_resources').declare_namespace(__name__)
> except ImportError:
>     import pkgutil
>     __path__ = pkgutil.extend_path(__path__, __name__)
> # pydevd_plugin_shape_str.py
> import sys
> sys.path.append(r'C:\Users\ruiqi\.cursor\extensions\ms-python.debugpy-2024.6.0-win32-x64\bundled\libs\debugpy\_vendored\pydevd')
> print(sys.path)
> 
> from _pydevd_bundle.pydevd_extension_api import StrPresentationProvider
> # from pydevd_helpers import find_mod_attr, find_class_name
> 
> def find_mod_attr(mod_name, attr):
>     mod = find_cached_module(mod_name)
>     if mod is None:
>         return None
>     return getattr(mod, attr, None)
> 
> 
> def find_class_name(val):
>     class_name = str(val.__class__)
>     if class_name.find(".") != -1:
>         class_name = class_name.split(".")[-1]
> 
>     elif class_name.find("'") != -1:  # does not have '.' (could be something like <type 'int'>)
>         class_name = class_name[class_name.index("'") + 1 :]
> 
>     if class_name.endswith("'>"):
>         class_name = class_name[:-2]
> 
>     return class_name
> 
> 
> class SizedShapeStr:
>     '''Displays the size of a Sized object before displaying its value.
>     '''
>     def can_provide(self, type_object, type_name):
>         sized_obj = find_mod_attr('collections.abc', 'Sized')
>         return sized_obj is not None and issubclass(type_object, sized_obj)
> 
>     def get_str(self, val):
>         if hasattr(val, 'shape'):
>             return f'shape: {val.shape}, value: {val}'
>         return f'len: {len(val)}, value: {val}'
> 
> if not sys.platform.startswith("java"):
>     StrPresentationProvider.register(SizedShapeStr)
> [pydevd_plugins.zip pydevd_plugins.zip](https://github.com/user-attachments/files/17478421/pydevd_plugins.zip)

Your directory structure should like this:

![Image](https://github.com/user-attachments/assets/0f84091a-ea8f-4248-b5e1-04bc510a82b5)

the extension directory is unnecessary in your zip file.

**Comment by joihn** [comment]
Quick workaround for this issue:
once your debugger has hit a breakpoint, input this in the debug console
```
import torch
def custom_repr(self):
    return f'{{Tensor:{tuple(self.shape)}}}'

original_repr = torch.Tensor.__repr__
torch.Tensor.__repr__ = custom_repr
```

![Image](https://github.com/user-attachments/assets/ea035e63-a9e7-4ae8-aa7d-d182964557eb)

**Comment by zhanghm1995** [comment]
@qige96 Hello, thanks for sharing this wonderful solution. It works well in my simple code project. However, when I want to use this solution in one of my big project that use **debugpy** and the **attach** method to debug my project, this solution can not work.

The example `main.py` code is:
```python
import torch
import numpy as np
import debugpy
debugpy.listen(("localhost", 5678))
print("Waiting for debugger attach")
debugpy.wait_for_client()
print("Debugger Attached!")

def custom_repr(self):
    return f'{{Tensor:{tuple(self.shape)}}}'

torch.Tensor.__repr__ = custom_repr


data1 = torch.randn(2, 3, 4)
print(data1.shape)

np_data = np.random.randn(2, 3)
print(np_data.shape)
```

And the `.vscode/launch.json` is:
```json
{
    "configurations": [

        {
            "name": "Python Debugger: Current File",
            "type": "debugpy",
            "request": "attach",
            "connect": {
                "host": "localhost",
                "port": 5678
            }
        }
    ]
}
```

So I wonder whether you have another solution to solve this problem?
